partialmessages: add per peer bounds on peer initiated groups

MarcoPolo opened the pull request:
A peer initiated group state is a state that is initiatlized due to receiving a partial message RPC for which we have not (yet) published to.

It's useful to have this state so we can remember what parts a peer has if we do publish to that group.

Previously we limited the total number of peer initiated group states in a topic. This was agnostic to how many groups a single peer initiated, which meant that a single peer could exhaust the whole limit.

This change tracks how many group states each peer has initialized and limits each peer to a configurable number of these states. The default total and per peer limits of 255 8 are likely good enough for most use cases, but the options exist for users to change this if they need to.

cc @kasey who suggested the idea.

MarcoPolo committed:
partialmessages: add per peer bounds on peer initiated groups

A peer initiated group state is a state that is initiatlized due to
receiving a partial message RPC for which we have not (yet) published
to.

It's useful to have this state so we can remember what parts a peer has
if we do publish to that group.

Previously we limited the total number of peer initiated group states in a
topic. This was agnostic to how many groups a single peer initiated,
which meant that a single peer could exhaust the whole limit.

This change tracks how many group states each peer has initialized and
limits each peer to a configurable number of these states. The default
total and per peer limits of 255 8 are likely good enough for most use
cases, but the options exist for users to change this if they need to.

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -111,13 +117,26 @@ type PartialMessageExtension struct {
 
-	peerInitiatedGroupCountPerTopic map[string]int
+	// map[topic]counter
+	peerInitiatedGroupCounter map[string]*peerInitiatedGroupCounterState
```
Reminder to fix the commented out code.

sukunrt commented on partialmessages/partialmsgs.go:
```diff
@@ -111,13 +117,26 @@ type PartialMessageExtension struct {
 	//     expected bounds?
 	ValidateRPC func(from peer.ID, rpc *pb.PartialMessagesExtension) error
 
+	// PeerInitiatedGroupLimitPerTopic limits the number of Group states all
+	// peers can initialize per topic. A group state is initialized by a peer if
+	// the peer's message marks the first time we've seen a group id.
+	PeerInitiatedGroupLimitPerTopic int
+
+	// PeerInitiatedGroupLimitPerTopicPerPeer limits the number of Group states
+	// a single peer can initialize per topic. A group state is initialized by a
+	// peer if the peer's message marks the first time we've seen a group id.
+	PeerInitiatedGroupLimitPerTopicPerPeer int
```
I'd recommend rephrasing the comments to talk of messages with group IDs as opposed to group state. The state is private to the package so the comment might be confusing. The clarifying statement helps, but if it's directly related to the number of new messages that'd be clearer.

MarcoPolo replied:
I'm not sure phrasing this in terms of messages is clearer. The thing this is limiting is the number of group ID related state for a group ID we don't know about.

Messages sounds like we are persisting application data, which we aren't.

Do you have a another suggestion?

MarcoPolo committed:
partialmessages: rename field to reflect use

changes partialMessageStatePerTopicGroup to
partialMessageStatePerGroupPerTopic to better reflect its usage.
